Assignment3/PointAndGrid.py

```diff
@@ -0,0 +1,280 @@
+            for c in range(C):
+                if self._grid[r][c] == element:
+                    result.append(Point(self, (r, c), element))
+        return result
+
+    def findNeighborsOfPoint(self, tuple):
+        """
+        Return list of Points that are neighbours of element with the given coordinates.
+        :param tuple: Tuple
+        :return: list(Point) or None
+        """
+        r = tuple[0]
+        c = tuple[1]
+        result = []
+        possible_coord = [(r-1, c-1),(r-1, c),(r, c-1),(r-1, c+1),
+                          (r+1, c-1),(r+1, c),(r, c+1),(r+1, c+1)]
+
+        for coord in possible_coord:
+            if self._validCoordinates(coord):
+                result.append(Point(self, coord, self.getElement(coord)))
+        return result if result else None
+
+    def ifWordInGrid(self, word):
+        """
+        Return True if the word can be formed in grid, False otherwise.
+        :param word: str
+        :return: bool
+        """
+        word = word.lower()
+        return self.findString(word) is not None
+
+    ############################## THE SOLUTION TO ASSIGNMENT 3 ##########################################
+
+    def findString(self, word):
+        """
+        Find the word in the grid where we can starting from one position move to any neighbors,
+        but can't visit the same cell twice. Return the path if exists, None otherwise.
+        :param word: string
+        :return: list(Point) or None
+        """
+        startPoints = self.findElementInGrid(word[0])
+        path = []
+        for startP in startPoints:
+            # we are here only if the first letter of word is in grid
+            path = [startP]
+            path = self._findPathRecursively(startP, word, path, [])
+        if self._pathIsWord(path, word):
+            return path
+
+    def _pathIsWord(self, path, word):
+        if not path:
+            return False
+        return [p.getValue() for p in path] == list(word)
+
+    def _findPathRecursively(self, point, word, path, visited):
+        """
+        Helper-function to find a word in a grid. Finds the longest path to form the word (may be only a part)
+        :param point: Point
+        :param word: str
+        :param path: list(Point)
+        :param visited: list(Point)
+        :return: list(Point) or None
+        """
+        if self._pathIsWord(path, word):
+            return path
+
+        possible_ways = self.findPossibleWays(path, point, word, visited)
+        if not possible_ways:
+            # we at the same time pop the last element from path and append it to list visited
+            visited.append(path.pop(-1))
+            if path:
+                self._findPathRecursively(path[-1], word, path, visited)
+
+        for neighbor in possible_ways:
+            path.append(neighbor)
+            self._findPathRecursively(neighbor, word, path, visited)
+
+        return path
+
+    def findPossibleWays(self, path, point, word, visited):

```

Comment:
Just to make the API cleander I'd suggest `path`, `point`, `word` and `visited` in the same order as they are in `_findPathRecursively()` signature.